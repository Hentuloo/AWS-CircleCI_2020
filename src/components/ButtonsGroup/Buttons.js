import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const Button = styled.button`
  position: absolute;
  max-width: 35%;
  border: none;
  padding: 5px 10px;
  cursor: pointer;
  display: block;
  background-color: transparent;
  transform: translate(-50%, 0%);
  &:focus {
    ${({ theme }) => theme.mediaQuery.md} {
      outline: 0;
    }
  }
`;
const Icon = styled.img`
  max-width: 100%;
  max-height: 100%;
`;
export const ButtonsWrapper = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0%;
  left: 0%;
  opacity: 0;
  pointer-events: none;
  transition: opacity 0.2s 0.15s linear;
  ${({ length, theme }) => {
    switch (length) {
      case 1:
        return css`
          ${Button}:nth-of-type(1) {
            left: 50%;
            top: 0%;
          }
        `;
      case 2:
        return css`
          ${Button}:nth-of-type(1) {
            left: 50%;
            top: 0%;
          }
          ${Button}:nth-of-type(2) {
            left: 50%;
            bottom: 1%;
          }
        `;
      case 3:
        return css`
          ${Button}:nth-of-type(1) {
            left: 50%;
            top: -1%;
          }
          ${theme.mediaQuery.md} {
            ${Button}:nth-of-type(1) {
              top: -4%;
            }
          }
          ${Button}:nth-of-type(2) {
            left: 22%;
            bottom: 4%;
          }
          ${theme.mediaQuery.md} {
            ${Button}:nth-of-type(2) {
              left: 20%;
            }
          }
          ${Button}:nth-of-type(3) {
            right: 22%;
            bottom: 4%;
            transform: translate(50%, 0%);
          }
          ${theme.mediaQuery.md} {
            ${Button}:nth-of-type(3) {
              right: 22%;
            }
          }
        `;
      case 4:
        return css`
          ${Button}:nth-of-type(1) {
            left: 25%;
            top: 5%;
          }
          ${Button}:nth-of-type(2) {
            right: 25%;
            top: 5%;
            transform: translate(50%, 0%);
          }
          ${Button}:nth-of-type(3) {
            left: 25%;
            bottom: 8%;
          }
          ${Button}:nth-of-type(4) {
            right: 25%;
            bottom: 8%;
            transform: translate(50%, 0%);
          }
        `;
      default:
        return null;
    }
  }}
`;

const Buttons = ({ buttons, onClick }) => {
  return (
    <ButtonsWrapper length={buttons.length}>
      {buttons.map(({ value, title, icon }, index) => (
        <Button
          type="button"
          key={value}
          title={title}
          onClick={() => onClick(index)}
        >
          <span className="sr-only">{title}</span>
          <Icon src={icon} alt={title} />
        </Button>
      ))}
    </ButtonsWrapper>
  );
};

Buttons.propTypes = {
  buttons: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      value: PropTypes.string,
      icon: PropTypes.string,
      onActive: PropTypes.node,
    }),
  ).isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Buttons;
