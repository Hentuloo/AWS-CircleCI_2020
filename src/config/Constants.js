export const Constants = {
  pl: {
    STATEMENTS: {
      buttonMoreInfoTitle: 'Więcej informacji',
      buttonMoreInfoAlt: 'Przycisk więcej informacji',
      findClient: 'Znajdź klienta...',
      sort: {
        alt: 'Sortowanie',
        email: 'Sortuj według alfabetu (email)',
        age: 'Sortuj według wieku',
        descending: 'malejąco',
        ascending: 'rosnąco',
      },
      sex: {
        alt: 'Płeć',
        male: 'Mężczyzna',
        female: 'Kobieta',
      },
      icons: {
        place: 'Miejsce zamieszkania',
        personIcon: 'avatar',
      },
    },
  },
};
