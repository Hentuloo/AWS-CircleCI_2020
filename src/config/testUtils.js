import React from 'react';
import PropTypes from 'prop-types';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import GlobalStyles from 'themes/GlobalStyles';
import theme from 'themes/mainTheme';

export const withTheme = (ui, { customTheme } = {}) => {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={customTheme || theme}>
      <GlobalStyles />
      <div>{children}</div>
    </ThemeProvider>
  );
  Wrapper.propTypes = {
    children: PropTypes.node.isRequired,
  };
  return {
    ...render(ui, { wrapper: Wrapper }),
  };
};
