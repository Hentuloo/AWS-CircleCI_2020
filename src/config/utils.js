export const sortByAlphabet = (array, parametr, reverse) => {
  if (reverse) {
    return array.sort((a, b) =>
      a[parametr].localeCompare(b[parametr]),
    );
  }
  return array.sort((a, b) => b[parametr].localeCompare(a[parametr]));
};

export const sortByNumber = (array, parametr, reverse) => {
  if (reverse) {
    return array.sort(
      (a, b) => parseFloat(b[parametr]) - parseFloat(a[parametr]),
    );
  }
  return array.sort(
    (a, b) => parseFloat(a[parametr]) - parseFloat(b[parametr]),
  );
};

export const isIncludeText = (text, filter) =>
  text.toLowerCase().includes(filter.toLowerCase());
