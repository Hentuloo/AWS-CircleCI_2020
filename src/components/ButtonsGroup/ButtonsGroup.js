import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Buttons, { ButtonsWrapper } from './Buttons';

const Wrapper = styled.div`
  position: relative;
  width: 130px;
  height: 130px;
  border-radius: 50%;
  margin: -10px auto;
  &:hover,
  &:active,
  &:focus {
    &::before {
      transform: scale(1.1);
    }
    &::after {
      transform: scale(0.46);
    }
    ${ButtonsWrapper} {
      pointer-events: auto;
      opacity: 1;
    }
  }

  &::before,
  &::after {
    position: absolute;
    content: '';
    width: 100%;
    height: 100%;
    left: 0%;
    top: 0%;
    border-radius: 50%;
    background-color: ${({ theme }) => theme.color.blue[2]};
    z-index: -1;
    transition: transform 0.4s ease;
    transform: scale(0.65);
  }
  &::after {
    background-color: ${({ theme }) => theme.color.blue[1]};
  }
`;

const InnerClircle = styled.div`
  position: absolute;
  left: 50%;
  top: 50%;
  font-size: ${({ theme }) => theme.fs.xs};
  transform: translate(-50%, -50%);
  color: white;
`;

const ButtonsGroup = ({
  buttons,
  className,
  name,
  onChange,
  title,
}) => {
  const [activeIndex, setActiveIndex] = useState(0);

  const handleButtonClick = buttonIndex => {
    setActiveIndex(buttonIndex);
    onChange({ target: { name, value: buttons[buttonIndex].value } });
  };
  if (buttons.length > 0) {
    return (
      <Wrapper className={className} tabindex="0">
        {title && <p className="sr-only">{title}</p>}
        <InnerClircle data-testid="onActive-text">
          {buttons[activeIndex].onActive}
        </InnerClircle>
        <Buttons buttons={buttons} onClick={handleButtonClick} />
      </Wrapper>
    );
  }
  return null;
};

ButtonsGroup.propTypes = {
  buttons: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      value: PropTypes.string,
      icon: PropTypes.string,
      onActive: PropTypes.node,
    }),
  ),
  className: PropTypes.string,
  name: PropTypes.string,
  onChange: PropTypes.func,
  title: PropTypes.string,
};

ButtonsGroup.defaultProps = {
  buttons: [],
  className: '',
  name: '',
  onChange: () => null,
  title: null,
};

export default ButtonsGroup;
