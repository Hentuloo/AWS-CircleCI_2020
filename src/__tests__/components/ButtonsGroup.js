import React from 'react';
import ButtonsGroup from 'components/ButtonsGroup/ButtonsGroup';
import { withTheme } from 'config/testUtils';
import { fireEvent } from '@testing-library/react';
import 'jest-styled-components';

const elementWithActiveText = 'onActive-text';

afterEach(() => {
  jest.clearAllMocks();
});

const defaultProps = {
  name: 'multiple_checkbox',
  onChange: jest.fn(),
  title: 'nice checkbox blah',
  buttons: [
    {
      value: 'some_default_value',
      title: 'some first value',
      icon: 'nice icon',
      onActive: 'my text from default button',
    },
    {
      value: 'first_value',
      title: 'add some sorter',
      icon: 'nice icon2',
      onActive: 'my text from first button',
    },
    {
      value: 'second_value',
      title: 'add some other sorter',
      icon: 'nice icon3',
      onActive: 'my text from second button',
    },
  ],
};

const setup = (props = defaultProps) => {
  const el = withTheme(<ButtonsGroup {...props} />);
  return {
    ...el,
  };
};

describe('ButtonsGroup component', () => {
  test('mocked buttons snapshot', () => {
    const { container } = setup();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should display title prop', () => {
    const { getByText } = setup();

    const titleOfCheckbox = getByText(defaultProps.title);

    expect(titleOfCheckbox).toBeInTheDocument();
  });

  test('should display options', () => {
    const { getByText } = setup();

    defaultProps.buttons.forEach(({ title }) => {
      expect(getByText(title)).toBeInTheDocument();
    });
  });

  test('should choose first button as default (as active)', () => {
    const { getByTestId } = setup();

    const activeText = getByTestId(elementWithActiveText);

    expect(activeText).toBeInTheDocument();
    expect(activeText).toHaveTextContent(
      defaultProps.buttons[0].onActive,
    );
    expect(activeText).not.toHaveTextContent('some other value');
  });

  test('should change active option by button click', () => {
    const { getAllByRole, getByTestId } = setup();

    const activeText = getByTestId(elementWithActiveText);
    const buttons = getAllByRole('button');

    expect(activeText).toHaveTextContent(
      defaultProps.buttons[0].onActive,
    );
    expect(buttons[1]).toHaveTextContent(
      defaultProps.buttons[1].title,
    );

    fireEvent.click(buttons[1]);

    expect(activeText).toHaveTextContent(
      defaultProps.buttons[1].onActive,
    );
  });

  test('should dispatch callback function from props', () => {
    const { getAllByRole, getByTestId } = setup();

    const buttons = getAllByRole('button');
    const activeText = getByTestId(elementWithActiveText);

    expect(defaultProps.onChange).toHaveBeenCalledTimes(0);

    fireEvent.click(buttons[2]);

    expect(activeText).toHaveTextContent(
      defaultProps.buttons[2].onActive,
    );
    expect(defaultProps.onChange).toHaveBeenCalledTimes(1);
    expect(defaultProps.onChange).toHaveBeenCalledWith({
      target: {
        name: defaultProps.name,
        value: defaultProps.buttons[2].value,
      },
    });
  });
});
