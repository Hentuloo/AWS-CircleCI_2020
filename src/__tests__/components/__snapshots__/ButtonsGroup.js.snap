// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonsGroup component mocked buttons snapshot 1`] = `
.c5 {
  position: absolute;
  max-width: 35%;
  border: none;
  padding: 5px 10px;
  cursor: pointer;
  display: block;
  background-color: transparent;
  -webkit-transform: translate(-50%,0%);
  -ms-transform: translate(-50%,0%);
  transform: translate(-50%,0%);
}

.c6 {
  max-width: 100%;
  max-height: 100%;
}

.c3 {
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0%;
  left: 0%;
  opacity: 0;
  pointer-events: none;
  -webkit-transition: opacity 0.2s 0.15s linear;
  transition: opacity 0.2s 0.15s linear;
}

.c3 .c4:nth-of-type(1) {
  left: 50%;
  top: -1%;
}

.c3 .c4:nth-of-type(2) {
  left: 22%;
  bottom: 4%;
}

.c3 .c4:nth-of-type(3) {
  right: 22%;
  bottom: 4%;
  -webkit-transform: translate(50%,0%);
  -ms-transform: translate(50%,0%);
  transform: translate(50%,0%);
}

.c0 {
  position: relative;
  width: 130px;
  height: 130px;
  border-radius: 50%;
  margin: -10px auto;
}

.c0:hover::before,
.c0:active::before,
.c0:focus::before {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c0:hover::after,
.c0:active::after,
.c0:focus::after {
  -webkit-transform: scale(0.46);
  -ms-transform: scale(0.46);
  transform: scale(0.46);
}

.c0:hover .c2,
.c0:active .c2,
.c0:focus .c2 {
  pointer-events: auto;
  opacity: 1;
}

.c0::before,
.c0::after {
  position: absolute;
  content: '';
  width: 100%;
  height: 100%;
  left: 0%;
  top: 0%;
  border-radius: 50%;
  background-color: rgb(116,171,179);
  z-index: -1;
  -webkit-transition: -webkit-transform 0.4s ease;
  -webkit-transition: transform 0.4s ease;
  transition: transform 0.4s ease;
  -webkit-transform: scale(0.65);
  -ms-transform: scale(0.65);
  transform: scale(0.65);
}

.c0::after {
  background-color: rgb(70,147,158);
}

.c1 {
  position: absolute;
  left: 50%;
  top: 50%;
  font-size: 1.35em;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  color: white;
}

@media (min-width:768px) {
  .c5:focus {
    outline: 0;
  }
}

@media (min-width:768px) {
  .c3 .c4:nth-of-type(1) {
    top: -4%;
  }
}

@media (min-width:768px) {
  .c3 .c4:nth-of-type(2) {
    left: 20%;
  }
}

@media (min-width:768px) {
  .c3 .c4:nth-of-type(3) {
    right: 22%;
  }
}

@media (min-width:768px) {
  .c0 {
    margin: 5px auto;
    width: 140px;
    height: 140px;
  }
}

@media (min-width:768px) {
  .c0:hover::before,
  .c0:active::before,
  .c0:focus::before {
    -webkit-transform: scale(1.2);
    -ms-transform: scale(1.2);
    transform: scale(1.2);
  }
}

@media (min-width:768px) {
  .c0:hover::after,
  .c0:active::after,
  .c0:focus::after {
    -webkit-transform: scale(0.5);
    -ms-transform: scale(0.5);
    transform: scale(0.5);
  }
}

@media (min-width:768px) {

}

<div>
  <div
    class="c0"
  >
    <p
      class="sr-only"
    >
      nice checkbox blah
    </p>
    <div
      class="c1"
      data-testid="onActive-text"
    >
      my text from default button
    </div>
    <div
      class="c2 c3"
    >
      <button
        class="c4 c5"
        title="some first value"
        type="button"
      >
        <span
          class="sr-only"
        >
          some first value
        </span>
        <img
          alt="some first value"
          class="c6"
          src="nice icon"
        />
      </button>
      <button
        class="c4 c5"
        title="add some sorter"
        type="button"
      >
        <span
          class="sr-only"
        >
          add some sorter
        </span>
        <img
          alt="add some sorter"
          class="c6"
          src="nice icon2"
        />
      </button>
      <button
        class="c4 c5"
        title="add some other sorter"
        type="button"
      >
        <span
          class="sr-only"
        >
          add some other sorter
        </span>
        <img
          alt="add some other sorter"
          class="c6"
          src="nice icon3"
        />
      </button>
    </div>
  </div>
</div>
`;
